freshmaker/pyxis.py

```diff
@@ -0,0 +1,261 @@
+
+    def get_digests_by_nvrs(self, nvrs, params=None):
+        """
+        Get images' digests(manifest_list_digest field) by their nvrs
+
+        :param list nvrs: list of nvrs of ContainerImages to query Pyxis
+        :param dict params: parameters to add to GET request
+        :return: set of digests
+        :rtype: set
+        """
+        if params is None:
+            request_params = {}
+        else:
+            request_params = copy.deepcopy(params)
+        request_params['include'] = ','.join([
+            'data.brew', 'data.repositories.manifest_list_digest'])
+
+        # get all manifest_list_digest of ContainerImages we got from Pyxis
+        digests = set()
+        for nvr in nvrs:
+            for image in self._pagination(f'images/nvr/{nvr}', request_params):
+                if image.get('repositories') and \
+                        'manifest_list_digest' in image['repositories'][0]:
+                    digests.add(image['repositories'][0]['manifest_list_digest'])
+
+        return digests
+
+    def filter_bundles_by_nvrs(self, nvrs, bundles):
+        """
+        Filter out bundles that don't have same digests in 'related_image' as
+        repositories.*.manifest_list_digest in images
+
+        :param list nvrs: nvrs to find images and filter bundles by them
+        :param list bundles: bundles to filter
+        :return: filtered list of bundles
+        :rtype: list
+        """
+        # get images by nvrs and then get their digests
+        image_digests = self.get_digests_by_nvrs(nvrs)
+
+        if not image_digests:
+            log.warning(f"The are no digests for nvrs: {','.join(nvrs)}")
+            return []
+
+        ret_bundles = []
+        # If bundle doesn't have any of digests of images, don't add it to return list
+        for bundle in bundles:
+            for image in bundle.get('related_images', []):
+                # If same digest within images' digests is found, it will be in return list
+                if image.get('digest') in image_digests:
+                    ret_bundles.append(bundle)
+                    break
+
+        return ret_bundles
+
+    def get_images_by_digests(self, digests, params=None):
+        """
+        Get bundle ContainerImages by the digests in their
+        'repositories.*.manifest_list_digest'
+
+        :param list digests: digests for Pyxis filter inside query
+        :param dict params: parameters to add to GET request
+        :return: dictionary with mapping digests to ContainerImage data
+        :rtype: dict
+        """
+        if params is None:
+            request_params = {}
+        else:
+            request_params = copy.deepcopy(params)
+        request_params['include'] = 'data.brew'
+        bundle_images = {}
+        for digest in digests:
+            # Filter by repositories.manifest_list_digest field
+            request_params['filter'] = \
+                f"repositories.manifest_list_digest=={digest}"
```

Comment:
Pyxis supports "or" filtering and I think it'd be more efficient than making a request for each digest.
For example:
```
https://pyxis.domain.local/v1/images?filter=repositories.manifest_list_digest==sha256:bacfe9f3b9e90de703682a75bcae4b88353c29831dd66294ecc501c6fb65c4ce,repositories.manifest_list_digest==sha256:1292722ec0827dc38b0af68350512803db94cffb7e9fc7dbb471ef8e95d11af8
```

You could potentially do this in batches of digests to reduce the filtering load on Pyxis if it's too slow.

Additionally, since you are just including "data.brew", you can just return a set of NVRs from this method rather than the dictionary you are building.

Reply:
Yes, and there is '=in=' operator for filtering so I can use it instead of a lot of ORs. 

About returning just NVRs. I thought about it but I thing for future that method can be extended to include other fields into the query and in that case returning list of images could be useful. So I would leave that as it is.